### PR TITLE
libyul as static library, eliminating circular dependencies

### PR DESCRIPTION
**Pull request by christianparpart**
# RFC

libyul as static library, eliminating circular dependencies

### NOTES

* This PR is based on my prior PR  (#5427) to factor out shared bits (error reporting, ...), which must be merged beforehand.
* I had to fully (and manually ;-( ) recreate this PR as I've been recreating #5427, too, which is a prerequisite of this PR.
* This PR deprecates #5399.

### CODE REVIEW

File: liblangutil/CharStream.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	size_type searchStart = min<size_type>(m_source.size(), _position);
+	if (searchStart > 0)
+		searchStart--;
+	size_type lineStart = m_source.rfind('\n', searchStart);
+	if (lineStart == string::npos)
+		lineStart = 0;
+	else
+		lineStart++;
+	return m_source.substr(lineStart, min(m_source.find('\n', lineStart),
+										  m_source.size()) - lineStart);
```
Comment (leonardoalt) [inline liblangutil/CharStream.cpp:89]: What happened to this line?

**Comment by chriseth** [comment]
Please rebase!

**Comment by axic** [comment]
@christianparpart I think this PR has similar issues, mixing a lot of changes into single commits :)

File: libyul/backends/evm/EVMCodeTransform.cpp [hunk 2]
```diff
@@ -29,10 +29,10 @@
 
 using namespace std;
 using namespace dev;
-using namespace dev::yul;
+using namespace yul;
 using namespace dev::solidity;
 
-using Scope = dev::solidity::assembly::Scope;
+// using Scope = yul::Scope;
```
Comment (axic) [inline libyul/backends/evm/EVMCodeTransform.cpp:35]: Do we need to keep the comment?

**Pull request closed without merging**